### PR TITLE
Restore support for libgcrypt

### PR DESCRIPTION
**Pull request by hg**
This is an attempt to restore support for libgcrypt as an alternative crypto backend.

By "99% done" I meant "I am not seeing any more issues", of course, so it's probably 50% done. Although it's been working fine for me for the past day, I'll be testing it more heavily tomorrow on real systems.

CI jobs now do an additional libgcrypt run, although not with sanitizers. If you run it locally with sanitizers, it will fail without changes from #324. I'll add libgcrypt support to sanitizer runs later, as it requires some relatively big changes.

I sliced it into separate commits for easier review. The meat is in 23096107dc2a95ef364e7a8ab851482734052f5f.

- https://github.com/hg/tinc/actions/runs/1140511821
- https://github.com/hg/tinc/actions/runs/1140348419
- https://github.com/hg/tinc/actions/runs/1138226100 (pretty much the same code, but with a different base64 encoder)

---

Just for fun, using default settings:

## libgcrypt

```
$ iperf3 -c 10.1.1.1
Connecting to host 10.1.1.1, port 5201
[  5] local 192.168.122.224 port 22580 connected to 10.1.1.1 port 5201
[ ID] Interval           Transfer     Bitrate         Retr  Cwnd
[  5]   0.00-1.00   sec  3.69 GBytes  31.7 Gbits/sec    0   3.00 MBytes
[  5]   1.00-2.00   sec  4.00 GBytes  34.4 Gbits/sec    0   3.00 MBytes
[  5]   2.00-3.00   sec  3.67 GBytes  31.5 Gbits/sec    0   3.00 MBytes
[  5]   3.00-4.00   sec  3.50 GBytes  30.0 Gbits/sec    0   3.00 MBytes
[  5]   4.00-5.00   sec  3.66 GBytes  31.3 Gbits/sec    0   3.00 MBytes
[  5]   5.00-6.00   sec  3.89 GBytes  33.4 Gbits/sec    0   3.00 MBytes
[  5]   6.00-7.00   sec  3.59 GBytes  30.9 Gbits/sec    0   3.00 MBytes
[  5]   7.00-8.00   sec  3.87 GBytes  33.2 Gbits/sec    0   3.00 MBytes
[  5]   8.00-9.00   sec  3.87 GBytes  33.2 Gbits/sec    0   3.00 MBytes
[  5]   9.00-10.00  sec  2.72 GBytes  23.4 Gbits/sec    0   3.00 MBytes
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-10.00  sec  36.4 GBytes  31.3 Gbits/sec    0             sender
[  5]   0.00-10.12  sec  36.4 GBytes  30.9 Gbits/sec                  receiver
```

## OpenSSL

```
$ iperf3 -c 10.1.1.1
Connecting to host 10.1.1.1, port 5201
[  5] local 192.168.122.224 port 56179 connected to 10.1.1.1 port 5201
[ ID] Interval           Transfer     Bitrate         Retr  Cwnd
[  5]   0.00-1.00   sec  3.94 GBytes  33.8 Gbits/sec    0   3.00 MBytes
[  5]   1.00-2.00   sec  3.86 GBytes  33.1 Gbits/sec    0   3.00 MBytes
[  5]   2.00-3.00   sec  3.88 GBytes  33.2 Gbits/sec    0   3.00 MBytes
[  5]   3.00-4.00   sec  3.93 GBytes  33.9 Gbits/sec    0   3.00 MBytes
[  5]   4.00-5.00   sec  3.90 GBytes  33.5 Gbits/sec    0   3.00 MBytes
[  5]   5.00-6.00   sec  3.91 GBytes  33.6 Gbits/sec    0   3.00 MBytes
[  5]   6.00-7.00   sec  3.94 GBytes  33.9 Gbits/sec    0   3.00 MBytes
[  5]   7.00-8.00   sec  3.86 GBytes  33.2 Gbits/sec    0   3.00 MBytes
[  5]   8.00-9.00   sec  3.90 GBytes  33.5 Gbits/sec    0   3.00 MBytes
[  5]   9.00-10.00  sec  3.88 GBytes  33.3 Gbits/sec    0   3.00 MBytes
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-10.00  sec  39.0 GBytes  33.5 Gbits/sec    0             sender
[  5]   0.00-10.12  sec  39.0 GBytes  33.1 Gbits/sec                  receiver
```

### CODE REVIEW

File: src/gcrypt/cipher.c [hunk 1]
```diff
@@ -31,25 +31,20 @@ static struct {
...
-	{NULL, GCRY_CIPHER_AES192, GCRY_CIPHER_MODE_ECB, 422},
-	{"aes192", GCRY_CIPHER_AES192, GCRY_CIPHER_MODE_CBC, 423},
-	{NULL, GCRY_CIPHER_AES192, GCRY_CIPHER_MODE_CFB, 425},
-	{NULL, GCRY_CIPHER_AES192, GCRY_CIPHER_MODE_OFB, 424},
-
-	{NULL, GCRY_CIPHER_AES256, GCRY_CIPHER_MODE_ECB, 426},
-	{"aes256", GCRY_CIPHER_AES256, GCRY_CIPHER_MODE_CBC, 427},
-	{NULL, GCRY_CIPHER_AES256, GCRY_CIPHER_MODE_CFB, 429},
-	{NULL, GCRY_CIPHER_AES256, GCRY_CIPHER_MODE_OFB, 428},
+	{"aes-128-ecb", GCRY_CIPHER_AES, GCRY_CIPHER_MODE_ECB, 418},
```
Comment (gsliepen) [inline src/gcrypt/cipher.c:34]: Why drop Blowfish? It was the default cipher in tinc 1.0 a few releases back.

File: src/gcrypt/rsagen.c [hunk 2]
```diff
@@ -20,220 +20,299 @@
...
+#include "rsa.h"
+#include "pem.h"
+
+// ASN.1 tags.
+typedef enum {
+	TAG_INTEGER = 2,
+	TAG_SEQUENCE = 16,
+} asn1_tag_t;
+
+static size_t der_tag_len(size_t n) {
```
Comment (gsliepen) [inline src/gcrypt/rsagen.c:36]: This is why I really dislike gcrypt. They don't support PEM, they have those stupid s-expressions... it doesn't help security if you force users to implement their own parsers for bog standard crypto file formats :(

Comment (gsliepen) [reply src/gcrypt/rsagen.c:36]: Oh I just saw that there are some companion libraries to libgcrypt, one of them is [libksba](https://gnupg.org/software/libksba/index.html), which might be able to do the PEM decoding.

File: src/gcrypt/cipher.c (hunk 1 above)
Comment (hg) [reply src/gcrypt/cipher.c:34]: It's not really considered secure anymore, as far as I understand. However, seeing that it impacts backwards compatibility, I'll return it (a bit later).

File: src/ed25519/ecdsagen.c [hunk 3]
```diff
@@ -54,7 +54,7 @@ static bool write_pem(FILE *fp, const char *type, void *vbuf, size_t size) {
 
 	while(size) {
 		size_t todo = size > 48 ? 48 : size;
-		b64encode(buf, base64, todo);
+		b64encode_tinc(buf, base64, todo);
```
Comment (gsliepen) [inline src/ed25519/ecdsagen.c:57]: Yes. When I wrote this code I don't even think there was a standardized way to encode Ed25519 keys into PEM format. Maybe it's better to remove the PEM encoding for Ed25519 keys completely; by default `tinc init` generates it as a regular config variable. I think only `tinc fsck` uses this functionality, and that can be changed.

File: src/gcrypt/cipher.c (hunk 1 above)
Comment (gsliepen) [reply src/gcrypt/cipher.c:39]: It's weak when using it to encrypt more than 4 GB of data if I understand it correctly. Tinc rotates keys regularly, but there might be an issue if you send that much in one go. But what's worse is the ECB mode support below, I think that should definitely be removed.

File: src/gcrypt/rsagen.c (hunk 2 above)
Comment (hg) [reply src/gcrypt/rsagen.c:36]: I don't think it helps with PEM encoding or decoding.

Its [reader](https://github.com/gpg/libksba/blob/88392670f4e5640744b8bd6a68debd2e05274d19/src/reader.c) and [writer](https://github.com/gpg/libksba/blob/88392670f4e5640744b8bd6a68debd2e05274d19/src/writer.c) are simple wrappers around `fread()` that don't do much (you still have to provide base64 encoding/decoding, write out PEM headers, etc. That's what GPG does: [read](https://github.com/gpg/gnupg/blob/283ccbc824d8062b86e27818a84ab5a29facd7a5/common/ksba-io-support.c#L177), [write](https://github.com/gpg/gnupg/blob/283ccbc824d8062b86e27818a84ab5a29facd7a5/common/ksba-io-support.c#L427).)

It seems to have a pretty decent DER builder. I almost finished porting the code and only then noticed that it's [less than 1.5 years](https://github.com/gpg/libksba/commit/cf49d3e60a67180fcb1b9005d910f015b388cf3c) old and is available on relatively few platforms right now:

- Debian stable
- Fedora (latest)
- Ubuntu (latest)
- FreeBSD 13
- OpenBSD 6.9
- NetBSD 9.2

but not:

- Debian oldstable
- Ubuntu 20.04 (current LTS)
- RHEL 8
- OpenBSD 6.8
- NetBSD 8.2

Any thoughs? Should I keep the hand-written DER encoder and use ksba through conditional compilation, or do we declare libgcrypt as supported only on recent systems?

Comment (gsliepen) [reply src/gcrypt/rsagen.c:36]: Hm, let's keep the hand-written code then, and move to the external library when it's more mature.

**Comment by hg** [comment]
Oops. I've been very busy for the past few days and haven't really managed to finish this yet. It works fine though (at least in my personal experience), and the remaining changes are not very important. I'll send them separately a bit later:

- update docs to mention libgcrypt as an alternative to openssl
- disable ECB modes (see diff below)
- a few other bits and pieces

I should have mentioned this here, sorry.

```diff
diff --git a/src/gcrypt/cipher.c b/src/gcrypt/cipher.c
index 0f7b008d..eb17f7d0 100644
--- a/src/gcrypt/cipher.c
+++ b/src/gcrypt/cipher.c
@@ -31,22 +31,18 @@ static struct {
 } ciphertable[] = {
        {"none", GCRY_CIPHER_NONE, GCRY_CIPHER_MODE_NONE, 0},

-       {NULL, GCRY_CIPHER_BLOWFISH, GCRY_CIPHER_MODE_ECB, 92},
        {"blowfish", GCRY_CIPHER_BLOWFISH, GCRY_CIPHER_MODE_CBC, 91},
        {NULL, GCRY_CIPHER_BLOWFISH, GCRY_CIPHER_MODE_CFB, 93},
        {NULL, GCRY_CIPHER_BLOWFISH, GCRY_CIPHER_MODE_OFB, 94},

-       {"aes-128-ecb", GCRY_CIPHER_AES, GCRY_CIPHER_MODE_ECB, 418},
        {"aes-128-cbc", GCRY_CIPHER_AES, GCRY_CIPHER_MODE_CBC, 419},
        {"aes-128-cfb", GCRY_CIPHER_AES, GCRY_CIPHER_MODE_CFB, 421},
        {"aes-128-ofb", GCRY_CIPHER_AES, GCRY_CIPHER_MODE_OFB, 420},

-       {"aes-192-ecb", GCRY_CIPHER_AES192, GCRY_CIPHER_MODE_ECB, 422},
        {"aes-192-cbc", GCRY_CIPHER_AES192, GCRY_CIPHER_MODE_CBC, 423},
        {"aes-192-cfb", GCRY_CIPHER_AES192, GCRY_CIPHER_MODE_CFB, 425},
        {"aes-192-ofb", GCRY_CIPHER_AES192, GCRY_CIPHER_MODE_OFB, 424},

-       {"aes-256-ecb", GCRY_CIPHER_AES256, GCRY_CIPHER_MODE_ECB, 426},
        {"aes-256-cbc", GCRY_CIPHER_AES256, GCRY_CIPHER_MODE_CBC, 427},
        {"aes-256-cfb", GCRY_CIPHER_AES256, GCRY_CIPHER_MODE_CFB, 429},
        {"aes-256-ofb", GCRY_CIPHER_AES256, GCRY_CIPHER_MODE_OFB, 428},
@@ -109,7 +105,7 @@ static bool cipher_open(cipher_t *cipher, int algo, int mode) {
        cipher->keylen = gcry_cipher_get_algo_keylen(algo);
        cipher->blklen = gcry_cipher_get_algo_blklen(algo);
        cipher->key = xmalloc(cipher->keylen + cipher->blklen);
-       cipher->padding = mode == GCRY_CIPHER_MODE_ECB || mode == GCRY_CIPHER_MODE_CBC;
+       cipher->padding = mode == GCRY_CIPHER_MODE_CBC;

        return true;
 }
```

**Pull request merged**